orca/service.go

```diff
@@ -0,0 +1,214 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package orca
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/orca/internal"
+	"google.golang.org/grpc/status"
+
+	v3orcapb "github.com/cncf/xds/go/xds/data/orca/v3"
+	v3orcaservicegrpc "github.com/cncf/xds/go/xds/service/orca/v3"
+	v3orcaservicepb "github.com/cncf/xds/go/xds/service/orca/v3"
+)
+
+func init() {
+	internal.AllowAnyMinReportingInterval = func(so *ServiceOptions) {
+		so.allowAnyMinReportingInterval = true
+	}
+}
+
+// minReportingInterval is the absolute minimum value supported for
+// out-of-band metrics reporting from the ORCA service implementation
+// provided by the orca package.
+const minReportingInterval = 30 * time.Second
+
+// Service provides an implementation of the OpenRcaService as defined in the
+// [ORCA] service protos.
+//
+// Server applications can use the SetXxx() and DeleteXxx() methods to record
+// measurements corresponding to backend metrics, which eventually get pushed to
+// clients who have initiated the SteamCoreMetrics streaming RPC.
+//
+// [ORCA]: https://github.com/cncf/xds/blob/main/xds/service/orca/v3/orca.proto
+type Service struct {
```

Comment:
We should document that this must be constructed via `NewService` or `Register`.  Why did we decide to keep `Register` instead of `NewService` actually?  `NewService` is more flexible, and `Register` is only a minor convenience for users who could call that one extra function themselves.  I'm fine with having both, but I believe `NewService` is what we should provide if we only have one.

Reply:
I think we decided to add `Register` to be like other services (eg: reflection, health etc). I have both now. 

I think `Register` is nice because the user will not have to depend directly on that proto. And also because the load report proto that they anyways have to handle is different from the service proto.

Reply:
> I think we decided to add Register to be like other services (eg: reflection, health etc). I have both now.

Looking back at the original comment, I was suggesting `Register` as the name instead of `EnableOutOfBandMetricsReporting`, but also to probably not include it in favor of providing the service implementation instead.  But I'm fine with having both.